aspnetcore/web-api/action-return-types/samples/3x/WebApiSample.Api.30/Controllers/ProductsController.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using WebApiSample.DataAccess.Models;
+using WebApiSample.DataAccess.Repositories;
+
+namespace WebApiSample.Api._30.Controllers
+{
+    [ApiController]
+    [Produces("application/json")]
+    [Route("[controller]")]
+    public class ProductsController : ControllerBase
+    {
+        private readonly ProductsRepository _repository;
+
+        public ProductsController(ProductsRepository repository)
+        {
+            _repository = repository;
+        }
+
+        #region snippet_Get
+        [HttpGet]
+        public IEnumerable<Product> Get() =>
+            _repository.GetProducts();
+        #endregion
+
+        #region snippet_GetById
+        [HttpGet("{id}")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        public ActionResult<Product> GetById(int id)
+        {
+            if (!_repository.TryGetProduct(id, out var product))
+            {
+                return NotFound();
+            }
+
+            return product;
+        }
+        #endregion
+
+        #region snippet_GetByPage
+        [HttpGet("{pageNumber:int:min(1)}/{pageSize:int:min(1)}")]
+        public async IAsyncEnumerable<Product> GetByPage(
+            int pageNumber,
+            int pageSize)
+        {
+            var products = _repository.GetProductsByPage(pageNumber, pageSize);
+            
+            await foreach (var product in products)
+            {
+                yield return product;
+            }
+        }
+        #endregion
+
+        #region snippet_GetByPageChunk
+        [HttpGet("chunk/{pageNumber:int:min(1)}/{pageSize:int:min(1)}")]
+        public async IAsyncEnumerable<Product> GetByPageChunk(
+            int pageNumber, 
+            int pageSize)
+        {
+            int firstRecord = pageNumber * pageSize;
+            int stopRecord = firstRecord + pageSize;
+
+            while (firstRecord < stopRecord)
+            {
+                var products = _repository.GetProductsByPage(pageNumber, pageSize);
```

Comment:
Since the 2nd arg is `pageSize`, I think this would always call `GetProductsByPage` exactly once. If `pageSize` was hard coded to `10`, as you mention in the text, this would behave as you described.

Reply:
I think I see the source of the confusion. I need to clarify in my explanation of the example. The `pageSize` parameter comes from the route parameter (10 in this case). The `pageNumber` parameter comes from the route parameter too (1 in this case).

Reply:
I really like the final sample. This is a perfect scenario for AsyncEnumerables.